File: orgo_mcp/tools/actions.py
"""Screen action tools — mouse, keyboard, and screenshot control."""

import base64

from mcp.server.fastmcp import Image

from orgo_mcp.server import mcp
from orgo_mcp.auth import get_current_api_key, resolve_computer_id
from orgo_mcp.client import computer_action
from orgo_mcp.errors import handle_orgo_error
from orgo_mcp.models import ComputerIdInput, ClickInput, TypeInput, KeyInput, ScrollInput, DragInput


@mcp.tool(
    name="orgo_screenshot",
    annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": False},
)
async def orgo_screenshot(params: ComputerIdInput) -> Image:
    """Take a screenshot of the full VM display (all windows, desktop). For Chrome browser tab only, prefer orgo_chrome_screenshot."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        data = await computer_action("GET", computer_id, "screenshot", api_key, direct=True)
        image_b64 = data.get("image", "")
        image_b64_clean = image_b64.strip().replace("\n", "").replace("\r", "").replace(" ", "")
        pad = len(image_b64_clean) % 4
        if pad:
            image_b64_clean += "=" * (4 - pad)
        return Image(data=base64.b64decode(image_b64_clean), format="png")
    except Exception as e:
        raise RuntimeError(handle_orgo_error(e))


@mcp.tool(
    name="orgo_click",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": False},
)
async def orgo_click(params: ClickInput) -> str:
    """Click at pixel (x, y) coordinates on the VM display. For Chrome elements, prefer orgo_chrome_click with element refs."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        payload = {"x": params.x, "y": params.y, "button": params.button}
        if params.double:
            payload["double"] = True
        await computer_action("POST", computer_id, "click", api_key, json=payload)
        action = "Double-clicked" if params.double else f"{'Right' if params.button == 'right' else ''}Clicked"
        return f"{action} at ({params.x}, {params.y})"
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_type",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": False},
)
async def orgo_type(params: TypeInput) -> str:
    """Type text at the current cursor position on the VM. For Chrome form fields, prefer orgo_chrome_form_input with element refs."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        await computer_action("POST", computer_id, "type", api_key, json={"text": params.text})
        preview = params.text[:50] + "..." if len(params.text) > 50 else params.text
        return f"Typed: {preview}"
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_key",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": False},
)
async def orgo_key(params: KeyInput) -> str:
    """Press a key or combo: Enter, Tab, Escape, ctrl+c, alt+Tab, ctrl+shift+s, F1-F12."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        await computer_action("POST", computer_id, "key", api_key, json={"key": params.key})
        return f"Pressed: {params.key}"
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_scroll",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": False},
)
async def orgo_scroll(params: ScrollInput) -> str:
    """Scroll the VM display up or down. For Chrome page scrolling, prefer orgo_chrome_scroll."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        await computer_action("POST", computer_id, "scroll", api_key,
                              json={"direction": params.direction, "amount": params.amount})
        return f"Scrolled {params.direction} by {params.amount}"
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_drag",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": False},
)
async def orgo_drag(params: DragInput) -> str:
    """Drag from (start_x, start_y) to (end_x, end_y) on the VM display. For drag-and-drop, text selection, window resizing in native apps."""
    try:
        api_key = get_current_api_key(mcp)
        computer_id = resolve_computer_id(params.computer_id)
        await computer_action("POST", computer_id, "drag", api_key, json={
            "start_x": params.start_x, "start_y": params.start_y,
            "end_x": params.end_x, "end_y": params.end_y,
            "button": "left", "duration": params.duration,
        })
        return f"Dragged ({params.start_x},{params.start_y}) -> ({params.end_x},{params.end_y})"
    except Exception as e:
        return handle_orgo_error(e)

File: src/tools/actions.ts
/**
 * Screen action tools -- mouse, keyboard, and screenshot control.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { getApiKey, resolveComputerId } from "../auth.js";
import { computerAction } from "../client.js";
import { handleError } from "../errors.js";
import type { ScreenshotResponse } from "../types.js";
import { registerOrgoTool } from "./registry.js";

// Model coordinate space -- matches orgo-web's CUA
const MODEL_WIDTH = 1280;
const MODEL_HEIGHT = 720;

// Cache actual VM resolutions: computer_id -> [width, height]
const vmResolutionCache = new Map<string, [number, number]>();

function scaleCoords(computerId: string, x: number, y: number): [number, number] {
  const vm = vmResolutionCache.get(computerId);
  if (!vm || (vm[0] === MODEL_WIDTH && vm[1] === MODEL_HEIGHT)) {
    return [x, y];
  }
  return [
    Math.round(x * (vm[0] / MODEL_WIDTH)),
    Math.round(y * (vm[1] / MODEL_HEIGHT)),
  ];
}

export function registerActionTools(server: McpServer): void {
  registerOrgoTool(server, {
    name: "orgo_screenshot",
    title: "Take Screenshot",
    description: "Take a screenshot of the full VM display (all windows, desktop). Returns an image.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: true,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const data = (await computerAction("GET", id, "screenshot", apiKey, {
          direct: true,
        })) as unknown as ScreenshotResponse;

        // Cache VM resolution for coordinate scaling
        if (data.width && data.height) {
          vmResolutionCache.set(id, [data.width, data.height]);
        }

        // Clean base64 image data
        let imageB64 = (data.image || "")
          .trim()
          .replace(/\n/g, "")
          .replace(/\r/g, "")
          .replace(/ /g, "");

        // Fix padding
        const pad = imageB64.length % 4;
        if (pad) imageB64 += "=".repeat(4 - pad);

        return {
          content: [
            {
              type: "image" as const,
              data: imageB64,
              mimeType: "image/png",
            },
          ],
        };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });

  registerOrgoTool(server, {
    name: "orgo_click",
    title: "Click",
    description: "Click at pixel (x, y) coordinates on the VM display. Coordinates are in 1280x720 model space.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      x: z.number().int().min(0).describe("X coordinate (pixels from left)"),
      y: z.number().int().min(0).describe("Y coordinate (pixels from top)"),
      button: z.enum(["left", "right"]).default("left").describe("Mouse button"),
      double: z.boolean().default(false).describe("Double-click if true"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id, x, y, button, double: dbl }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const [realX, realY] = scaleCoords(id, x, y);
        const payload: Record<string, unknown> = { x: realX, y: realY, button };
        if (dbl) payload.double = true;
        await computerAction("POST", id, "click", apiKey, { json: payload });
        const action = dbl ? "Double-clicked" : button === "right" ? "Right-clicked" : "Clicked";
        return { content: [{ type: "text" as const, text: `${action} at (${x}, ${y})` }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });

  registerOrgoTool(server, {
    name: "orgo_type",
    title: "Type Text",
    description: "Type text at the current cursor position on the VM.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      text: z.string().min(1).describe("Text to type at cursor position"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id, text }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        await computerAction("POST", id, "type", apiKey, { json: { text } });
        const preview = text.length > 50 ? text.substring(0, 50) + "..." : text;
        return { content: [{ type: "text" as const, text: `Typed: ${preview}` }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });

  registerOrgoTool(server, {
    name: "orgo_key",
    title: "Press Key",
    description: "Press a key or combo: Enter, Tab, Escape, ctrl+c, alt+Tab, ctrl+shift+s, F1-F12.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      key: z.string().min(1).describe("Key or combo: Enter, Tab, Escape, ctrl+c, alt+Tab"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id, key }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        await computerAction("POST", id, "key", apiKey, { json: { key } });
        return { content: [{ type: "text" as const, text: `Pressed: ${key}` }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });

  registerOrgoTool(server, {
    name: "orgo_scroll",
    title: "Scroll",
    description: "Scroll the VM display up or down.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      direction: z.enum(["up", "down"]).describe("Scroll direction"),
      amount: z.number().int().min(1).max(20).default(3).describe("Scroll clicks (1-20)"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id, direction, amount }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        await computerAction("POST", id, "scroll", apiKey, { json: { direction, amount } });
        return { content: [{ type: "text" as const, text: `Scrolled ${direction} by ${amount}` }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });

  registerOrgoTool(server, {
    name: "orgo_drag",
    title: "Drag",
    description: "Drag from (start_x, start_y) to (end_x, end_y). Coordinates in 1280x720 model space.",
    inputSchema: {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      start_x: z.number().int().min(0).describe("Start X coordinate"),
      start_y: z.number().int().min(0).describe("Start Y coordinate"),
      end_x: z.number().int().min(0).describe("End X coordinate"),
      end_y: z.number().int().min(0).describe("End Y coordinate"),
      duration: z.number().min(0.1).max(5.0).default(0.5).describe("Drag duration in seconds"),
    },
    toolsets: ["screen"],
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
    handler: async ({ computer_id, start_x, start_y, end_x, end_y, duration }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const [realSx, realSy] = scaleCoords(id, start_x, start_y);
        const [realEx, realEy] = scaleCoords(id, end_x, end_y);
        await computerAction("POST", id, "drag", apiKey, {
          json: {
            start_x: realSx,
            start_y: realSy,
            end_x: realEx,
            end_y: realEy,
            button: "left",
            duration,
          },
        });
        return {
          content: [{ type: "text" as const, text: `Dragged (${start_x},${start_y}) -> (${end_x},${end_y})` }],
        };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    },
  });
}
